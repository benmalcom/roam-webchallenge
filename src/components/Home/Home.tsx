import React, { Component, FormEvent } from 'react';
import './Home.scss';
import { connect } from 'react-redux';
import autobind from 'autobind-decorator';
import { getLocation } from '../../redux/actions/location';

interface ComponentProps {
	getLocation: Function;
	loading: boolean;
	location: object;
}

class Home extends Component<ComponentProps, any> {
	constructor(props) {
		super(props);
		this.state = {
			inputValue: '',
		};
	}

	@autobind
	getLocationDetails(e: FormEvent<HTMLButtonElement>) {
		const {inputValue} = this.state;
		const {getLocation} = this.props;
		if (inputValue) {
			getLocation(inputValue);
		}
	}

	@autobind
	onSearchChange({currentTarget: {name, value}}: FormEvent<HTMLInputElement>) {
		this.setState((state) => ({[name]: value}));
	}

	render() {
		const {loading} = this.props;
		const {inputValue} = this.state;
		return (<div className="container-fluid top">
			<div className="row col-12 col-sm-8 col-md-7 col-lg-7 mx-auto">
				<div className="text-center">
					<h4><strong>Find favourite spots in Helsinki</strong></h4>
					<p className="text-muted">Enter a location in Helsinki and it should show you the nearest
						amenities (stops, bike rentals, bike parking, and car parking)</p>
					<div className="input-group">
						<input className="form-control border-secondary py-2" name="inputValue"
							   onChange={this.onSearchChange} value={inputValue}
							   type="search" placeholder="Search a location..."/>
						<div className="input-group-append">
							<button className="btn btn-outline-info" disabled={loading}
									type="button" onClick={this.getLocationDetails}>
								<i className={loading ? 'fa fa-spinner fa-spin' : 'fa fa-search'}/>
							</button>
						</div>
					</div>
				</div>
			</div>
		</div>);
	}
}

const stateProps = ({uiReducer: {loading}}) => ({loading});
const dispatchProps = (dispatch) => ({
	getLocation: (name: string) => dispatch(getLocation(name)),
});
export default connect(stateProps, dispatchProps)(Home);
