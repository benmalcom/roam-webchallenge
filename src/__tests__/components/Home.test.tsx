import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import Home from '../../components/Home/Home';
import { Provider } from 'react-redux';

describe('<Home />', () => {
	const mockStore = configureStore();
	const initialState = {location: null, error: null};
	let store;
	let props;
	let component;
	beforeEach(() => {
		store = mockStore(initialState);
		props = {
			getLocation: jest.fn(() => 'getLocation'),
			loading: false,
			location: null,
		};
		component = shallow(<Provider store={store}><Home {...props} /></Provider>);
		component.setState({inputValue: ''});

	});

	test('Component renders without crash', () => {
		expect(component).toMatchSnapshot();
	});
	test('Component renders with initial props', () => {
		expect(component.props().getLocation).toBeDefined();
		expect(component.props().loading).toBeFalsy();
		expect(component.props().location).toBeNull();
	});
});
