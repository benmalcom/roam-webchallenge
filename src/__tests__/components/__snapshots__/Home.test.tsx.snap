// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Component renders without crash 1`] = `
<Connect(Home)
  getLocation={[MockFunction]}
  loading={false}
  location={null}
/>
`;
